linesearch.go

```diff
@@ -63,46 +60,34 @@ func (ls *LinesearchMethod) Iterate(loc *Location) (Operation, error) {
 		// Init was not called. What to do? What about ls's internal state?
 
 	case MajorIteration:
-		// We previously requested MajorIteration but since we're here, the
+		// We previously commanded MajorIteration but since we're here, the
 		// previous location was not good enough to converge the full
 		// optimization. Start the next linesearch and store the next
 		// evaluation point in loc.X.
-		return ls.initNextLinesearch(loc.X)
+		return ls.initNextLinesearch(loc)
 
 	default:
-		// Store the result of the previously requested evaluation into ls.loc.
-		if ls.lastOp&FuncEvaluation != 0 {
-			ls.loc.F = loc.F
-		}
-		if ls.lastOp&GradEvaluation != 0 {
-			copy(ls.loc.Gradient, loc.Gradient)
-		}
-		if ls.lastOp&HessEvaluation != 0 {
-			ls.loc.Hessian.CopySym(loc.Hessian)
-		}
-		// Update the indicator of valid fields of ls.loc.
+		// Update the indicator of valid fields of loc.
 		ls.eval |= ls.lastOp
 
 		if ls.nextMajor {
 			ls.nextMajor = false
 
 			// Linesearcher previously indicated that it had finished, but we
-			// needed to evaluate invalid fields of ls.loc. Now we have them and
+			// needed to evaluate invalid fields of loc. Now we have them and
 			// can announce MajorIteration.
 
-			copyLocation(loc, &ls.loc)
 			ls.lastOp = MajorIteration
 			return ls.lastOp, nil
 		}
 	}
 
-	projGrad := floats.Dot(ls.loc.Gradient, ls.dir)
-	if ls.Linesearcher.Finished(ls.loc.F, projGrad) {
-		// Form an operation that evaluates invalid fields of ls.loc.
-		ls.lastOp = complementEval(&ls.loc, ls.eval)
+	projGrad := floats.Dot(loc.Gradient, ls.dir)
```

Comment:
I planned to update this part properly when removing Finished from Linesearcher but I will include it here.